### PR TITLE
276 error using boosted 4 tooltip in an angular project

### LINKED ISSUE #358: Typedoc warnings for `ODSChartsCSSThemes` and `ODSChartsPopoverManagers`
### Description

After having [updated `typedoc` from 0.26.6 to 0.26.7](https://github.com/Orange-OpenSource/ods-charts/pull/357), running `npm run serve` raises 2 warning messages:

```
[warning] Failed to resolve link to "ODSChartsCSSThemes.NONE" in comment for ODSChartsThemeOptions.cssTheme
[warning] Failed to resolve link to "ODSChartsPopoverManagers.NONE" in comment for ODSChartsTheme.externalizePopover
```

These two elements have something in common: they're not enums but const with a key/value approach.

### PR DESCRIPTION
**Pull request by jacques-lebourgeois**
### Related issues

[Error using boosted 4 tooltip in an Angular projec](https://github.com/Orange-OpenSource/ods-charts/issues/276)
Closes #358.

### Description

I just pushed two commits:
- One for the bug: the bug isn't really solved here, but worked around. I don't know how to use the boosted global object when the packaging system used seems to make the boosted variable inaccessible. I'm interested in solutions. In the meantime, I'm doing a fallback on the popover and tooltip renderer using Apache Echarts customized to the ODS charter.
- The second commit is an attempt to improve the documentation to make it more readable and fluid.

### Motivation & Context

Cf bug and its comments

### Types of change

- Bug fix (non-breaking which fixes an issue)
- Doc enhancement

### Test checklist

Please check that the following tests projects are still working:

- [ ] `docs/examples`
- [ ] `test/angular-ngx-echarts`
- [ ] `test/angular-echarts`
- [ ] `test/html`
- [ ] `test/react`
- [ ] `test/vue`
- [ ] `test/examples/bar-line-chart`
- [ ] `test/examples/single-line-chart`
- [ ] `test/examples/timeseries-chart`

### CODE REVIEW

File: src/theme/ods-chart-theme.ts [hunk 1]
```diff
@@ -553,16 +577,12 @@ export class ODSChartsTheme {
...
-   *
-   *   {@link ODSChartsPopoverManagers} gives preconfigured renderer
-   *
-   *   default value is  {@link ODSChartsPopoverManagers.NONE}: uses default Apache ECharts template to externalize tooltip/popover HTML element, implementing Orange Design system
-   *
-   * optionally you can use this call to set dataOptions
-   * @returns returns back the theme manager object
+   * @param popoverConfig the configuration of the externalizePopover feature {@link ODSChartsPopoverConfig}
+   * @param popoverDefinition renderer {@link ODSChartsPopoverDefinition} of the popover/tooltip. **{@link ODSChartsPopoverManagers}** is used to
+   * specify preconfigured renderer for Apache ECharts, Boosted 5 or Boosted 4 renderer.
```
Comment (jacques-lebourgeois) [inline src/theme/ods-chart-theme.ts:582]: Add a warning that using Boosted 4 or Boosted 5 rendering requires dependency on the boosted library and access to the boosted global variable.

File: src/theme/popover/ods-chart-popover-definitions.ts [hunk 2]
```diff
@@ -268,10 +304,12 @@ class BOOSTED4_Definition extends ODSChartsPopoverDefinitionWithRenderer {
...
 /**
+ * `ODSChartsPopoverManagers` is used to specify popoverDefinition parameter (popup/tooltip renderer) of {@link ODSChartsTheme.externalizePopover}.
+ *
  * Available popover/tooltip renderer:
- * - ODSChartsPopoverManagers.NONE: to use default ECharts template to externalize tooltip/popover HTML element, implementing Orange Design System
- * - ODSChartsPopoverManagers.BOOSTED5: to use Boosted 5 tooltip/popover
- * - ODSChartsPopoverManagers.BOOSTED4: to use Boosted 4 tooltip/popover
+ * - `ODSChartsPopoverManagers.NONE`: to use default ECharts template to externalize tooltip/popover HTML element, implementing Orange Design System
+ * - `ODSChartsPopoverManagers.BOOSTED5`: to use Boosted 5 tooltip/popover
+ * - `ODSChartsPopoverManagers.BOOSTED4`: to use Boosted 4 tooltip/popover
```
Comment (jacques-lebourgeois) [inline src/theme/popover/ods-chart-popover-definitions.ts:312]: Add a warning that using Boosted 4 or Boosted 5 rendering requires dependency on the boosted library and access to the boosted global variable.

**Comment by jacques-lebourgeois** [comment]
I think this PR corrects the bug as expressed. 

I've added a warning about the dependency on the Boosted library for Popover and Tooltip rendering.

File: src/theme/ods-chart-theme.ts [hunk 3]
```diff
@@ -529,12 +551,14 @@ export class ODSChartsTheme {
...
    *
-   * externalizeLegends() needs:
-   * - echart: the initialized eCharts object
-   * - legendHolderSelector: the CSS selector of the legends container if it a string. But it can also be a {@link ODSChartsLegendHolderDefinition} or an array of {@link ODSChartsLegendHolderDefinition}
-   *
-   * optionally you can use this call to set dataOptions
-   * @returns returns back the theme manager object
+   * @param echart the initialized eCharts object
+   * @param legendHolderSelector
+   * legendHolderSelectorcan be
```
Comment (louismaximepiton) [inline src/theme/ods-chart-theme.ts:556]:
```suggestion
   * legendHolderSelector can be:
```

File: src/theme/ods-chart-theme.ts [hunk 4]
```diff
@@ -529,12 +551,14 @@ export class ODSChartsTheme {
...
+   * - a string, and then is interprated as the CSS selector of the legends container
```
Comment (louismaximepiton) [inline src/theme/ods-chart-theme.ts:557]:
```suggestion
   * - a string, and then is interpreted as the CSS selector of the legends container
```

File: src/theme/ods-chart-theme.ts [hunk 5]
```diff
@@ -553,16 +577,15 @@ export class ODSChartsTheme {
...
    * The generated tooltips or popover will implement the Orange Design System.
    *
-   * externalizePopover() needs:
-   * - popoverConfig: the configuration of the externalizePopover feature {@link ODSChartsPopoverConfig}
-   * - popoverDefinition: a renderer {@link ODSChartsPopoverDefinition} of the popover/tooltip
-   *
-   *   {@link ODSChartsPopoverManagers} gives preconfigured renderer
+   * @param popoverConfig the configuration of the externalizePopover feature {@link ODSChartsPopoverConfig}
+   * @param popoverDefinition renderer {@link ODSChartsPopoverDefinition} of the popover/tooltip. **{@link ODSChartsPopoverManagers}** is used to
+   * specify preconfigured renderer for Apache ECharts, Boosted 5 or Boosted 4 renderer.
```
Comment (louismaximepiton) [inline src/theme/ods-chart-theme.ts:582]:
```suggestion
   * specify preconfigured renderer for Apache ECharts, Boosted 5 or Boosted 4.
```

File: src/theme/ods-chart-theme.ts [hunk 6]
```diff
@@ -553,16 +577,15 @@ export class ODSChartsTheme {
...
    *
-   *   default value is  {@link ODSChartsPopoverManagers.NONE}: uses default Apache ECharts template to externalize tooltip/popover HTML element, implementing Orange Design system
+   * Default value is  {@link ODSChartsPopoverManagers.NONE}, that means it uses default Apache ECharts template to externalize tooltip/popover HTML element, implementing Orange Design system.
```
Comment (louismaximepiton) [inline src/theme/ods-chart-theme.ts:584]:
```suggestion
   * Default value is {@link ODSChartsPopoverManagers}.NONE, that means it uses default Apache ECharts template to externalize tooltip/popover HTML element, implementing Orange Design system.
```
The link seems to be broken, not sure my workaround works in here.

File: src/theme/ods-chart-theme.ts [hunk 7]
```diff
@@ -595,14 +616,14 @@ export class ODSChartsTheme {
...
+   * - options implementing the Orange Design System {@link ODSChartsThemeOptions}
    * - optionally options implementing {@link externalizeLegends},
    * - optionally options implementing {@link externalizePopover},
    * - optionally options implementing {@link manageChartResize},
    * - data from {@link setDataOptions}
    *
-   * optionally you can use this call to set dataOptions
-   * @returns the Apache ECharts options
+   * @param dataOptions optionally you can use this call to set dataOptions, if not already set.
+   * @returns the Apache ECharts options to use in [Apache Echarts setOption()](https://echarts.apache.org/en/option.html) call.
```
Comment (louismaximepiton) [inline src/theme/ods-chart-theme.ts:626]:
```suggestion
   * @returns the Apache ECharts options to use in [Apache Echarts `setOption()`](https://echarts.apache.org/en/option.html) call.
```

**Comment by louismaximepiton** [comment]
If we keep the new example directory, we probably want to update `dependabot.yml` to group update dependencies in this directory ?

**Comment by jacques-lebourgeois** [comment]
> Do we agree that none of the files in `./test/angular-ng-boosted` was changed ?

I'm not sure I understand. `./test/angular-ng-boosted` is a new test project to reproduce the bug. Do you want this test project to be uncommitted (removed from github)?

**Comment by jacques-lebourgeois** [comment]
> If we keep the new example directory, we probably want to update `dependabot.yml` to group update dependencies in this directory ?

Added

File: package.json [hunk 8]
```diff
@@ -23,7 +23,7 @@
     "serve:any": "serve",
     "serve:react": "serve test/react/build/",
     "serve:vue": "serve test/vue/dist/",
-    "typedoc": "typedoc --out docs/api src/ods-charts.ts --disableSources --excludePrivate --excludeProtected --readme docs/README.md && node build/add_head.mjs",
+    "typedoc": "typedoc --out docs/api src/ods-charts.ts --skipErrorChecking --disableSources --excludePrivate --excludeProtected --readme docs/README.md && node build/add_head.mjs",
```
[the remark this answers is not in the document]
Comment (jacques-lebourgeois) [reply package.json]: I removed it

File: .github/dependabot.yml [hunk 9]
```diff
@@ -43,6 +43,17 @@ updates:
...
           - "*"
+  - package-ecosystem: npm
+    directory: "/test/angular-ng-boosted"
+    labels:
+      - dependencies
+    schedule:
+      interval: weekly
+    versioning-strategy: increase
+    groups:
+      angular-echarts-all-deps:
```
Comment (louismaximepiton) [inline .github/dependabot.yml:54]:
```suggestion
      angular-ng-boosted-all-deps:
```

**Review by louismaximepiton — COMMENTED** [review]
Still some comments not taken into account. 

> I'm not sure I understand. ./test/angular-ng-boosted is a new test project to reproduce the bug. Do you want this test project to be uncommitted (removed from github)?

Nope, it looks a copy paste of a starter project, Just wanted if any file was changed compared to this starter

File: src/ods-charts.ts [hunk 10]
```diff
@@ -9,13 +9,9 @@
...
  *
- * Once the library integrated, via
- * - script &lt;script src="...ods-charts.js"&gt;
- * - or an import * as ODSCharts from 'ods-charts'
+ * This page is an index of the module's classes, interfaces and enumeration.
  *
- * use `ODSCharts.getThemeManager`({@link ODSChartsThemeOptions}) to get the {@link ODSChartsTheme}.
- *
- * The {@link ODSChartsTheme} is used to build the ODS theme and the charts options.
+ * Their use is introduced on the [home page](./#md:documentation).
```
[the remark this answers is not in the document]
Comment (jacques-lebourgeois) [reply src/ods-charts.ts:14]: `md:documentation` est l'ancre dans la page

et j'utilise un chemin relatif `./`pour que ça fonctionne en local

File: src/theme/ods-chart-theme.ts [hunk 11]
```diff
@@ -141,6 +141,9 @@ export enum ODSChartsMode {
   DARK = 'dark',
 }
 
+/**
+ * `ODSChartsThemeOptions` is used to set options of the theme [ODSCharts.getThemeManager( ODSChartsThemeOptions)](../#md:documentation).
```
[the remark this answers is not in the document]
Comment (jacques-lebourgeois) [reply src/theme/ods-chart-theme.ts:145]: Same reason, `md:documentation` is the anchor

../ is the parent page

**Comment by jacques-lebourgeois** [comment]
> Still some comments not taken into account.
> 

Sorry, Hope it is done now

> > I'm not sure I understand. ./test/angular-ng-boosted is a new test project to reproduce the bug. Do you want this test project to be uncommitted (removed from github)?
> 
> Nope, it looks a copy paste of a starter project, Just wanted if any file was changed compared to this starter

Oh sorry, I understand now :)

Sure, I made somes changes:
- remove js boosted inportation 
  - to reduce boosted importation to its importation from packaging and reproduce the bug
  - ie in angular.json, no more "node_modules/boosted/dist/js/boosted.bundle.js" 
- remove concequently some obsoletes codes
  - boosted-components that used boosted directly from js
  - routes to removed pages in app.routes.ts ,
- change configuration of externalized Popover to force usage of Boosted in ods-charts.component
  - replace `externalizePopover()` call by `externalizePopover({}, ODSCharts.ODSChartsPopoverManagers.BOOSTED5)`

File: src/ods-charts.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (louismaximepiton) [reply src/ods-charts.ts:14]: Ok nvm, it works well as-is, It's only 2 more warnings in the typedoc command, let's create an issue about this.

**Review by louismaximepiton — APPROVED** [review]
Looks good, thanks ❤️ Only a conflict to resolve and then :shipit:

**Pull request merged**